### DISCUSSION TITLE
Add an ESP32-C6 as a "coprocessor" and radio module (making PCB have both a ESP32-S3 and a ESP32-C6) so that second SoC can be used as Thread Border Router

### DISCUSSION THREAD

**Discussion by Hedda** [FutureProofHomes/Satellite1-Hardware#22] [Q&A] [answered]
@FutureProofHomes @that1guy @alextrical please consider also adding an ESP32-C6 (or ESP32-H2) as a "coprocessor" module as well to the same FutureProofHomes Satellite1 Development Board, i.e. add a second radio SoC so PCB will have both a ESP32-S3 as well as a ESP32-C6/ESP32-H2.

The main real-world use case reason to add an [ESP32-C6](https://www.espressif.com/en/products/socs/esp32-c6) (or an [ESP32-H2](https://www.espressif.com/en/products/socs/esp32-h2)) module other than just using it as a generic-coprocessor MCU SoC that could be used to offload stuff that ESP32-C6 (and ESP32-H2) has an IEEE 802.15.4 radio which means it can be used as a "Thread Border Router" (with OpenThread Border Router firmware), for a Thread network used by Matter standard fabric:

* https://www.home-assistant.io/integrations/matter/

  * https://www.home-assistant.io/integrations/thread/

If you add this and get that working then you should be able to add Thread's Border Router icon for marketing too:

* https://csa-iot.org/wp-content/uploads/2023/08/Matter-Retail-Education-Presenation.pdf

![image](https://github.com/user-attachments/assets/3a464c36-de74-452e-a050-aece7d5fca62)

Adding an extra ESP32 (ESP32-C6 or ESP32-H2) does take some space on PCB however should not add that much larger BOM cost,  but the additional possibilities such an extra ESP32 SoC could add should hopefully more than make up for that slightly higher cost!

You can possibly in the future alternativly also use ESP32-C6 (or ESP32-H2) on a single of them as a remote Zigbee Coordinator (also known as a Serial-over-IP Zigbee controller adapter) for [Home Assistant's built-in ZHA integration (native Zigbee Gateway)](https://www.home-assistant.io/integrations/zha), see/follow this work-in-progress but note that the ESP Zigbee radio for zigpy is is still very experimental and not yet fully working with ZHA:

* https://github.com/lhespress/zigpy-espzb
  * https://github.com/zigpy/zigpy-cli/pull/45
    * https://github.com/zigpy/zigpy/issues/1052

Another option would be to also use the ESP32-C6 as a "coprocessor" module a a dedicated Bluetooth Proxy chip with ESPHome:

* https://esphome.io/components/bluetooth_proxy.html

Note that the reason main you want to combine ESP32-S3 and ESP32-C6 is not only to off-load those processes but more importantly so you separate Wi-Fi signals from Thread/Zigbee signals, and that is due to WiFi tendacy to overwhelm and interfer with Thread/Zigbee communication because how they both work.

* https://www.home-assistant.io/integrations/zha#zigbee-interference-avoidance-and-network-rangecoverage-optimization

Same goes by the way for WiFi and Bluetooth Low Entergy but it usually is not a great impact on user experience as for Thread/Zigbee:

* https://www.home-assistant.io/integrations/bluetooth#bluetooth-interference-with-other-devices

Anyway, what I think what would make a product like FutureProofHomes Satellite1 perfect as a "Thread Border Router" is that a user can add multiple Thread Border Routers (TBR/OTBR) to their Thread/Matter network for redundacy (using fail-over) and a voice assistant is something that many Home Assistant users are likely to have many of and placed all over their home, so having many Thread Border Routers is one of the keys to getting a more robust Thread network.

* https://www.home-assistant.io/integrations/thread/#about-thread-border-routers

![image](https://github.com/user-attachments/assets/9b4150b4-848f-421f-a029-c59eae760c5a)

As proof-of-concept of a ESP32-S3 + ESP32-H2 combination board made for this purpose check out see the "ESP Thread Border Router" board which is Espressif official SDK reference hardware for this purpose (can use used as either Thread Border Router or Zigbee Coordinator):

* https://openthread.io/guides/border-router/espressif-esp32

Development is primarily being done with Espressif's ESP Thread Border Router SDK hardware (based on ESP32-H2-MINI-1 module):

- https://docs.espressif.com/projects/esp-thread-br/en/latest/
  - https://github.com/espressif/esp-zigbee-sdk/tree/main/examples/esp_zigbee_gateway
    - https://github.com/espressif/esp-thread-br
    - https://github.com/espressif/esp-idf/tree/master/examples/openthread/ot_br
    - https://openthread.io/guides/border-router/espressif-esp32

![image](https://github.com/user-attachments/assets/34c7079f-ce51-4946-a8aa-465c5ce7c0f1)

![image](https://github.com/user-attachments/assets/a40bdf65-d7ce-4b6e-a965-0862d7a6e05c)

OpenThread Border Router Example as in a similar two-SoC setup combining both a ESP32-S3 and ESP32-H2 in the same solution:

* https://github.com/espressif/esp-idf/tree/master/examples/openthread/ot_br

![image](https://github.com/user-attachments/assets/3186a2c2-9691-45c4-9f06-5e1ccc24347e)

**Comment by Hedda** [comment]
Maybe this should be moved to idea under discussions instead? Or? https://github.com/orgs/FutureProofHomes/discussions

**Comment by that1guy** [comment]
This is a very interesting FR and something we have discussed @Hedda.  

As of this moment, in order to get to beta launch quickly, we will not implement this (yet) --however I want to leave this ticket open as we may be willing to implement this in future revisions of the Sat1.

**Comment by Hedda** [comment]
> As of this moment, in order to get to beta launch quickly, we will not implement this (yet) --however I want to leave this ticket open as we may be willing to implement this in future revisions of the Sat1.

I suspect Home Assistant's first-revision devkit won't not have it either, but could read into schematics about things to come 😜

* https://www.home-assistant.io/integrations/thread/#about-thread-border-routers

![image](https://github.com/user-attachments/assets/77182e39-490a-44c8-82fe-7711f6fdd316)

_Image taken from [the Thread Smart Home Fact Sheet](https://www.threadgroup.org/support#Resources) by the Thread Group. It illustrates the landscape of Matter, Thread, and Border routers. Instead of Matter, you could also see another protocol here, such as HomeKit._

_Unlike other IoT protocols, Thread can use multiple border routers in a single network. This increases wireless coverage and reduces the risk of a single point of failure. Ideal for home automation, with a potentially large number of devices spread over a large area._

Slightly different image here on the threadgroup.org when they describe a Thread based Matter fabric using the same text but also mentioning that _Thread Border Router can be built int many devices such as access points, **smart speakers**, etc._:

* https://www.threadgroup.org/BUILT-FOR-IOT/Smart-Home

![image](https://github.com/user-attachments/assets/8e940dd3-67ba-4079-aa29-576214b2c280)

PS: Also check out this blog which better describe what a Thread Border Router actually is and why it is very good to have many:

* https://www.threadgroup.org/news-events/blog/ID/356/WHAT-IS-A-THREAD-BORDER-ROUTER-AND-HOW-IS-IT-DIFFERENT-FROM-A-HUB-OR-BRIDGE

![image](https://github.com/user-attachments/assets/d44bee7b-f12a-41c4-903a-9a22687379ee)

**Comment by Hedda** [comment]
@that1guy Oh, and in the mean-time it might be a good idea for you to order Espressif's ESP Thread Border Router / Zigbee GW:

* https://www.sparkfun.com/products/25497 

* https://www.aliexpress.com/item/1005005688193617.html

Suposedly you can get it working with Home Assistant:

* https://www.youtube.com/watch?v=nHH7j--lfDo

* https://github.com/home-assistant-libs/python-otbr-api/issues/61

**Comment by StuartIanNaylor** [comment]
I am not sure about the current ESP design or add another, as a thread border router would be great, but is the esp platform a good fit for that initial voice pipeline.
Apart wondering if esphome and microwakeword understand a few 101s of farfield and RiR's one of the ways to greatly improve a KW system is to learn from usage data. https://github.com/kahrendt/microWakeWord/issues/28#issuecomment-2564400870 as think the dataset creation and methods is something kahrendt has inherited from openwakeword.
https://ai.google.dev/edge/litert/models/ondevice_training provides a great way to do 'fine-tuning' but does tf4micro support those entry points.
So its great the hat potentially can sit on maybe something more substantial such as an Arm board. As with tflite all the models from https://github.com/google-research/google-research/tree/master/kws_streaming can be used and not just the one that kahrendt has done a great job untangling.
On ESP you don't have do ondevice training but on KW hit have a rolling window that will capture the KW and forward up to a device that can finetune the KWS with actual usage data. (It learns through use) and through OTA or other method update the esp model.
If tf4micro supports this or there is even the space to add rolling window capture methods I dunno, someone with esp knowledge would have to say yeah or nay. But that fine tuning on actual use data with local private training is the definitive method, even when you do actually start to create datasets of real use.

Going one step further with current advise of a minimum of a N100 is really due to using a humongeous LLM based ASR such as whisper and this really weird situation where we have to wait for a permissive licenced piece of opensource to be refactored and rebranded as the HA ASR, whilst there are many to choose from where we can credit the authors, use and feedback to the project, in a more opensource friendly manner.
If you get rid of whisper and don't shackle yourself to an LLM just to do simple intents, that a much lighter NLP toolkit would do as equally well, then a Pi5 or RK3588 board and less becomes possible.

So 2 questions here.
Is it possible to use the xmos voice processor as a USB device than a HAT? KWS even with lighter ASR than whisper are still a fraction of the process a ASR would take. On a device such as a Pi5 or N100 you could have several KWS running provided extended coverage of a zone?
The 2nd question also with current design is it possible to have active microphones on a daughter board to make isolation and positioning easy than the problems of a fixed board and hat? Even to create a balanced line driver where a small unobtrusive microphone daughter board could be some distance from the xmos board.

https://openthread.io/guides/border-router/raspberry-pi

**Comment by Hedda** [comment]
@that1guy @alextrical Another reason to have another ESP32 on the same board as a co-processor is it could handle a display, or?

As a proof-of-concept, do you think it would be possible to use the FutureProofHomes Satellite1 Dev Kit (or alternativly the [“ReSpeaker Lite 2-Mic Array” USB board with or without ESP32](https://www.seeedstudio.com/ReSpeaker-Lite-p-5928.html)) in combination with one of these 7-inch series of capacitive touchscreen displays from Elecrow in order to make a variant of a Home Assistant Voice Satellite with a screen?  ..i.e. something similar in design to Google Nest Hub and Amazon Echo Show 😉 

* https://www.cnx-software.com/2025/01/17/crowpanel-advance-esp32-s3-displays-with-replaceable-wifi-6-thread-zigbee-lora-and-2-4ghz-wireless-modules/

At least these "CrowPanel Advance" display series features "replaceable wireless modules" meaning that the ESP32 module can be replaced. That is, this article says that they feature an ESP32-S3-WROOM-1-N16R8 module soldered on the back + having a pin header taking ESP32-H2, ESP32-C6 or...

![Image](https://github.com/user-attachments/assets/308a06ca-7285-4650-8b40-3d1211fc3e64)

The specificiation does say that it also features a built-in microphone (however I assume that is rubbish) as well as a speaker connector (but no mention of a DAC).


The also have loads of different models with various sizes of screens. With the "Advanced" series all having a pre-soldered ESP32-S3-WROOM-1-N16R8 module:

![Image](https://github.com/user-attachments/assets/f2ebf275-62ec-423c-8b43-78ea5397ea7f)

CrowPanel Advance ESP32 displays are available on Aliexpress from $25.99 up to $54.19 depending on the selected model and whether any of the wireless modules shown above is added as an option. But you can also purchase the displays and wireless modules on Elecrow's company’s store.

* https://www.cnx-software.com/2025/01/17/crowpanel-advance-esp32-s3-displays-with-replaceable-wifi-6-thread-zigbee-lora-and-2-4ghz-wireless-modules/

PS: @StuartIanNaylor All of what you posted above about wake-words and using FutureProofHomes Satellite1 as a USB-microphone or design is totally off-topic here (same as in the forum), please start a new discussion instead of trying to high-jack others:

* https://github.com/FutureProofHomes/Satellite1-Hardware/discussions

**Comment by StuartIanNaylor** [comment]
@Hedda https://github.com/orgs/FutureProofHomes/discussions/9

**Comment by Hedda** [comment]
> please consider also adding an ESP32-C6 (or ESP32-H2) as a "coprocessor" module 

FYI, YouTuber "GreatScott!" showed of DIY Zigbee End Device with ESP32-C6 and ESP32-H2 in some proof-of-concept setups and what is interesting related to this request is that he tested range of them and found that the ESP32-C6 had much greater range:

https://www.youtube.com/watch?v=niqtYxXy6KQ

So that is another reason why you woud want to prefer the ESP32-C6 over the ESP32-H2.


> Note that the reason main you want to combine ESP32-S3 and ESP32-C6 is not only to off-load those processes but more importantly so you separate Wi-Fi signals from Thread/Zigbee signals, and that is due to WiFi tendacy to overwhelm and interfer with Thread/Zigbee communication because how they both work.
> 
> * https://www.home-assistant.io/integrations/zha#zigbee-interference-avoidance-and-network-rangecoverage-optimization
> 
> Same goes by the way for WiFi and Bluetooth Low Entergy but it usually is not a great impact on user experience as for Thread/Zigbee:
> 
> * https://www.home-assistant.io/integrations/bluetooth#bluetooth-interference-with-other-devices


Another FYI; YouTuber "Andreas Spiess" has show of prototype of new/upcoming [ESP32-C5 dev kit/board (ESP32-C5-DevKitC-1)](https://docs.espressif.com/projects/esp-dev-kits/en/latest/esp32c5/esp32-c5-devkitc-1/index.html) which also have a Zigbee/Thread radio, though in that video he only tested the 5GHz Wi-Fi capability of the ESP32-C5:

https://www.youtube.com/watch?v=glEvHRHdboA

So what makes the ESP32-C5 SoC interesting related to this request is that it could allow you to both act as a Thread Border Router (or Zigbee Coordinator) and a Wi-Fi module at the same time as long as configure it to only use 5GHz for WiFi (if Thread/Zigbee is in use).

**Comment by Hedda** [comment]
> You can possibly in the future alternativly also use ESP32-C6 (or ESP32-H2) on a single of them as a remote Zigbee Coordinator (also known as a Serial-over-IP Zigbee controller adapter) for [Home Assistant's built-in ZHA integration (native Zigbee Gateway)](https://www.home-assistant.io/integrations/zha), see/follow this work-in-progress but note that the ESP Zigbee radio for zigpy is is still very experimental and not yet fully working with ZHA:
> 
> * https://github.com/lhespress/zigpy-espzb
>   
>   * [Add Espressif System Zigbee radio library support zigpy/zigpy-cli#45](https://github.com/zigpy/zigpy-cli/pull/45)
>     
>     * [[REQUEST] zigpy-espzb (Espressif Zigbee NCP Serial Protocol) radio library for use with esp-zigbee-ncp onESP32-C6 and ESP32-H2 SOCs? zigpy/zigpy#1052](https://github.com/zigpy/zigpy/issues/1052)

FYI, regarding possible using such a radio module as a remote Zigbee Coordoinator, in addition to the above zigpy-espzb project there are now aboth ZHA/zigpy and Zigbee2MQTT/zigbee-herdsman developers are now working on totally independent separate projects that share the same goal of using a remote Zigbee Coordinator radio as a raw transceiver (transmitter and reciever) frontend via OpenThread RCP firmware and its Spinel protocol. Very early and only experimental but if interested then check out:

* https://github.com/zigpy/ziggurat/

and

* https://github.com/Nerivec/zigbee-on-host

or

* https://github.com/puddly/zigpy-spinel

Those solutions would not only allow running the actual Zigbee stack remotley on a host computer but also enable you to work around the main downside of Zigbee, which is that the physical Zigbee Coordionator adapter is a single-point-of-failure, as with this solution you can have a failover setup that makes it switch to a different remote radio adapter if one goes down (i.e. any of the Home Assistant Voice Sattelite with this could act as a remote Zigbee Coodrinator adapter and if you have many then you have redundancy).

In theory all that is needed on the ESP32-C6 side is running OpenThread RCP (Radio Co-Processor) firmware and that is available:

* https://github.com/espressif/esp-idf/blob/master/examples/openthread/ot_rcp/README.md
  * https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-guides/openthread.html
    * https://openthread.io/guides/border-router/espressif-esp32

**Comment by that1guy** [comment]
@Hedda I don't want to close issues that have such great knowledge in them.  This is a wonderful Github discussion and something we're discussing in the future.  Could you be so kind to move this knowledge and conversation into a Github discussion and then close this ticket?  Once we agree on a path forward in the Github Discussions we will create the actionable tickets and implement the solution.

**Comment by Hedda** [comment]
> [@Hedda](https://github.com/Hedda) … Could you be so kind to move this knowledge and conversation into a Github discussion and then close this ticket? Once we agree on a path forward in the Github Discussions we will create the actionable tickets and implement the solution.

I would if I could but you have move it yourself since only you as repo admin have permissions/rights to do so 😉

**Reply by that1guy** [reply]
DONE!  Apologies.  Thought community had access to do this.

**Comment by Hedda** [comment]
> As a proof-of-concept, do you think it would be possible to use the FutureProofHomes Satellite1 Dev Kit (or alternativly the [“ReSpeaker Lite 2-Mic Array” USB board with or without ESP32](https://www.seeedstudio.com/ReSpeaker-Lite-p-5928.html)) in combination with one of these 7-inch series of capacitive touchscreen displays from Elecrow in order to make a variant of a Home Assistant Voice Satellite with a screen? ..i.e. something similar in design to Google Nest Hub and Amazon Echo Show 😉
> 
> * https://www.cnx-software.com/2025/01/17/crowpanel-advance-esp32-s3-displays-with-replaceable-wifi-6-thread-zigbee-lora-and-2-4ghz-wireless-modules/

Off-topic but FYI; diybateriaslifepo4 managed to get C7277 Display Driver working in ESPHome with ["CrowPanel Advance" 7-inch model](https://www.aliexpress.com/item/1005008297553830.html) now:

* https://github.com/diybateriaslifepo4/Crowpanel-Advance-7

![image](https://github.com/user-attachments/assets/9a95673a-8feb-4706-817f-02f10caf7298)

PS: The datasheets are available in Elecrow-RD repository (and there is some demo code in the same repo):

https://github.com/Elecrow-RD/CrowPanel-Advance-HMI-ESP32-AI-Display/blob/master/7.0/screen_datasheet/

**Comment by Hedda** [comment]
Maybe could add a Grove port to the next hardware revision as that could enable users to play devices like this new M5Stack Unit Gateway H2?

* https://shop.m5stack.com/products/esp32-h2-thread-zigbee-gateway-unit

* https://docs.m5stack.com/en/unit/Unit%20Gateway%20H2

* https://www.youtube.com/watch?v=wjfyz15jWmw&ab_channel=M5Stack

FYI, [Home Assistant Voice Preview Edition](https://www.home-assistant.io/blog/2024/12/19/voice-preview-edition-the-era-of-open-voice/) added a Grove post since it allows easy plug-and-play to test sensors with ESPHome.

* https://voice-pe.home-assistant.io/guides/grove_port/

* https://wiki.seeedstudio.com/Grove_System/

M5Stack’s new Unit Gateway H2 is sold for ~$7 as a development kit with a Grove port and powered by a ESP32-H2-MINI-1 module:

* https://www.cnx-software.com/2025/03/17/m5stack-unit-gateway-h2-tiny-esp32-h2-thread-zigbee-gateway/

![Image](https://github.com/user-attachments/assets/73e154f7-6683-4529-b48a-094cd7c82564)

![Image](https://github.com/user-attachments/assets/e97d9555-e3dc-4cbd-8912-f02936825dfe)

![Image](https://github.com/user-attachments/assets/61bb2864-16b9-4075-8a5b-1f6e451c3a1a)

**Comment by that1guy** [comment, marked as answer]
To keep BOM costs down we're going to avoid adding another processor.  We will look into supporting thread via a accessory that attaches to the HAT's expansion port.  
![image](https://github.com/user-attachments/assets/b72ef182-caa3-4006-b52d-b7bb2c748d58)

**Comment by Hedda** [comment]
FYI, Guition P4 Series product models with both SoC modules as well as development boards and display boards now also feature their new Guition JC-ESP32P4-M3-C6 CoreBoard module (esp32-p4-display-module) which feature integrated ESP32-P4 and ESP32-C6 MCUs inside a single SoC module:

- https://www.guition.com/esp32p4-display-module/esp32p4-module

- https://www.guition.com/esp32p4-display-module/esp32-p4-display-module

  - https://www.guition.com/esp32p4-display-module/esp32p4-display-module

- https://www.cnx-software.com/2025/07/18/14-development-board-features-guition-esp32-p4-esp32-c6-module/

  - https://www.cnx-software.com/2025/08/12/4-3-inch-touch-display-board-features-single-esp32-p4-esp32-c6-module-supports-camera-and-speakers/

- https://www.guition.com/esp32p4-display-module/raspberry-pi-display-module

  - https://www.cnx-software.com/2025/05/06/7-inch-esp32-p4-wireless-touchscreen-display-supports-guition-designer-for-lvgl-arduino-and-esp-idf-programming/

- https://www.surenoo.com/products/27872757

- https://www.youtube.com/watch?v=KLwfSvutmaI

- https://devices.esphome.io/devices/guition-esp32-p4-m3-dev/

- https://github.com/jtenniswood/esphome-lvgl/tree/main/guition-esp32-p4-jc1060p470

<img width="700" height="700" alt="Image" src="https://github.com/user-attachments/assets/bc6ed107-d71c-4442-98e8-67f9161ca187" />

![Image](https://github.com/user-attachments/assets/0c5dc083-57ce-4d0d-98e7-850719157dbf)

<img width="1290" height="1249" alt="Image" src="https://github.com/user-attachments/assets/da4bd9cd-8914-468b-ba05-ec502b1a090e" />

<img width="1200" height="800" alt="Image" src="https://github.com/user-attachments/assets/71bd2506-911f-4d41-a037-9c5317c71de2" />
